docs/rfcs/0003-ecommerce-models.md

```diff
@@ -0,0 +1,136 @@
+## Title for RFC
+
+ecommerce workflow
+
+#### Abstract
+
+This describes the various states of ecommerce models and where and when we can and cannot
+modify data.
+
+#### Product
+
+A `Product` has a `price`, `description`, and a `GenericForeignKey` `content_object` which can
+refer to any Django model instance. An admin user should be able to edit
+`description` or `price` without affecting any previous purchase, so we need to also store
+this separately when the purchase is made.
+
+#### Basket
+
+A `Basket` has foreign keys to an `Order` and a `User`. The `User` foreign key must always exist.
+
+A `Basket` may be editable or not editable, depending on the `editable` boolean field. In general
+a `Basket` starts off editable so that the user can add products or a coupon. Then it becomes not
+editable once it is passed to CyberSource for payment.
+
+When a `Basket` is editable it should not have a link to an `Order`, and when it is not editable
+it should have a link to an `Order`. This makes the `editable` flag a little redundant but this
+makes the intent more explicit, and it provides flexibility in the future if we want to change
+the logic around when `Order` is created and linked to a `Basket`.
+
+There should be at most one `Basket` for a user which is editable.
+This is the user's active basket. If there is no editable `Basket` for the user,
+it should be created once the user adds a product or a coupon.
+
+A basket may have at most one coupon. See `CouponBasket` below.
+
+##### Edge cases
+
+Once a user checks out and is directed to CyberSource, the basket stops being editable even if
+they haven't actually paid yet. If the user tries to go back they will find their basket to be empty
+and they will have to add their items again.
+
+Fixing this complicates the logic significantly, and the UI does not really have a basket, so I'm
+ignoring this edge case for now.
+
+#### Line
+
+A `Line` is an item in a `Basket`. It has a foreign key to the `Basket` and to `Product`. There
+are also fields for `price_per_unit`, `total_price`, `quantity` and `description`.
```

Comment:
I'm not exactly sure what `price_per_unit` and `total_price` means for a Line. Is that to suppose purchasing multiples of the same course/product?

Related, we'd like to be very explicit about the price for reporting purposes. One proposal is to capture the "list_price" and the "price_paid" for each line in the order.

Reply:
To capture a conversation with @maxliu : the list_price should probably be copied from the product into the line at the time of purchase, in case the product price ever changes.